### PR TITLE
[Core] Add support for sampling penalties to v1 ngram speculative decoding

### PR DESCRIPTION
**Pull request by pooyadavoodi**
* Adds support for sampling parameter penalties to ngram speculative decoding
* Adds unit tests
* Adds a new triton kernel for expand_batch_to_tokens to support 2D tensors.

### CODE REVIEW

**Comment by linglingfancodeopensource** [comment]
interesting work!

**Comment by linglingfancodeopensource** [comment]
interesting work!

**Comment by pooyadavoodi** [comment]
@WoosukKwon @simon-mo @robertgshaw2-redhat I would appreciate a review of this PR.

File: vllm/v1/sample/rejection_sampler.py [hunk 1]
```diff
@@ -288,25 +290,71 @@ def compute_probs(
...
+    x: torch.Tensor,
+    cu_num_tokens: torch.Tensor,
     num_tokens: int,
     replace_from: int = 0,
     replace_to: int = 0,
 ) -> torch.Tensor:
-    """Expand [batch_size] tensor to [num_tokens] tensor based on the number of
-    tokens per batch in cu_num_tokens.
+    """Expand [batch_size] tensor to [num_tokens] tensor
+    (or [batch_size, vocab_size] to [num_tokens, vocab_size]) based on
```
Comment (WoosukKwon) [inline vllm/v1/sample/rejection_sampler.py:350]: I actually don't understand in which case we have this 2D tensor. Could you please elaborate more?

File: vllm/v1/sample/rejection_sampler.py [hunk 2]
```diff
@@ -318,18 +366,39 @@ def expand_batch_to_tokens(
...
+        expand_kernel_2d[(batch_size, vocab_size)](
+            expanded_x,
+            x,
+            cu_num_tokens,
+            replace_from,
+            replace_to,
+            MAX_NUM_TOKENS=MAX_SPEC_LEN,
+            VOCAB_SIZE=vocab_size,
+            num_warps=4 if vocab_size > 512 else 1,
+        )
```
Comment (WoosukKwon) [inline vllm/v1/sample/rejection_sampler.py:397]: If I understand correctly, this kernel is not vectorized enough, and therefore uses too many thread blocks. Please fix the kernel to do vectorized load. If it is implemented correctly, we should launch either `(batch_size,)` or `(batch_size, cdiv(vocab_size, BLOCK_SIZE)` thread blocks (depending on how it is implemented).

File: vllm/v1/sample/rejection_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/rejection_sampler.py:350]: Is the second dim `vocab_size`? Shouldn't it be something like `seq_len`?

File: vllm/v1/sample/rejection_sampler.py [hunk 3]
```diff
@@ -288,25 +290,71 @@ def compute_probs(
...
+            cu_num_draft_tokens,
+            num_tokens,
+        )
+        output_tokens_t = _convert_to_tensors(
+            sampling_metadata.output_token_ids, vocab_size, logits.device)
+        output_tokens_t = expand_batch_to_tokens(
+            output_tokens_t,
+            cu_num_draft_tokens,
+            num_tokens,
+        )
```
Comment (WoosukKwon) [inline vllm/v1/sample/rejection_sampler.py:311]: IIUC, we need different "output_token_ids` per draft token position, since the 1st draft token id should be counted for penalties in sampling the 2nd draft token, while the reverse doesn't hold. Is this considered in this PR?

**Review by WoosukKwon — COMMENTED** [review]
Hi @pooyadavoodi, thanks for submitting the PR and sorry for the delays in review.
Unfortunately, I don't quite understand the PR. Please check out my comments.
